src/util.h

```diff
@@ -175,4 +175,12 @@ static SECP256K1_INLINE void memczero(void *s, size_t len, int flag) {
     }
 }
 
+/** If flag is true, set *r equal to *a; otherwise leave it. Constant-time.  Both *r and *a must be initialized.*/
+static SECP256K1_INLINE void secp256k1_int_cmov(int *r, const int *a, int flag) {
+    uint32_t mask0, mask1;
+    mask0 = flag + ~((uint32_t)0);
```

Comment:
```suggestion
    unsigned int mask0, mask1;
    mask0 = flag + ~((unsigned int)0);
```
or shorter
```suggestion
    unsigned int mask0, mask1;
    mask0 = flag + ~0u;
```
is probably cleaner to review and more portable because `unsigned int` is guaranteed to have the same bitlength as `int`.

Reply:
I wonder now if I'm sure this function is 100% safe for signed integers (I'll need to look again at the integer promotion rules etc)
I think it is safe though?
(flag will probably be promoted to unsigned, and then the overflow will be safe and I think all AND operations on ints are safe )

Reply:
I thought about this too.

> (flag will probably be promoted to unsigned, and then the overflow will be safe and I think all AND operations on ints are safe )

Indeed, this was my understanding, too! (`flag` will not be "promoted" it will be "converted" to unsigned but this is just wording). The same is true in `(*r & mask0)` where `*r` will be converted to unsigned, and the same for `(*a & mask1)`, so everything on the right-hand side of the final assignment is unsigned arithmetic.

Reply:
Would it make sense to test this function with `a` equal `INT_MIN` and `INT_MAX`?

Reply:
I guess it can't hurt and it's cheap to do.

Reply:
I'll add a test, FYI passing some flags to clang shows us exactly how the conversions happen here: https://godbolt.org/z/t9GHKK

Though I wouldn't mind changing it to something like this: https://godbolt.org/z/5NxyE5

Reply:
Yeah explicit conversions would be nice. PR looks good. Thanks for the tests.

Reply:
argh, this cmov is implementation defined for any negative value :/
See http://port70.net/~nsz/c/c89/c89-draft.html#3.2.1.2
> an unsigned integer is converted to its corresponding signed integer, if the value cannot be represented the result is implementation-defined. 

So even `(int)(unsigned int)(-1)` is implementation defined behavior.

Reply:
Yeah, even though all non-obscure C implementations will do the most efficient thing possible, i.e., make the cast a noop on the bit level, which implies that casting back and forth is fine. So it will probably never be a problem for us. But it's cleaner not to rely on it. I suggest we simply `VERIFY` that inputs are nonnegative and a comment why this VERIFY makes sense?

Reply:
Yeah, people on #C in freenode told me I'm being overly paranoid lol, VERIFY is more than enough